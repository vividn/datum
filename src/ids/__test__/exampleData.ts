import { DatumData, DatumMetadata } from "../../documentControl/DatumDocument";

export const exampleData: DatumData = {
  foo: "abc",
  bar: "def",
  complex: { data: "nested" },
  array: ["various", 2, "data"],
  num: 3,
  "wei%rd": "da%ta",
};
export const exampleOccurTime = "2020-11-09T00:35:10.000Z";
export const exampleDataOccur = {
  occurTime: exampleOccurTime,
  occurUtcOffset: 1,
  ...exampleData,
};
export const exampleDataOccurField: DatumData = {
  ...exampleDataOccur,
  field: "main",
};
export const exampleMeta: DatumMetadata = {
  createTime: "2020-11-09T00:40:12.544Z",
  modifyTime: "2020-11-09T00:40:12.544Z",
  random: 0.7368733800261729,
  humanId: "mqp4znq4cvp3qnj74fgi9",
};
