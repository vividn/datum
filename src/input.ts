import yargs from "yargs";

export type DatumYargsType = {
  db?: string;
  host?: string;
  username?: string;
  password?: string;
  env?: string;
  baseData?: string;
  date?: string;
  yesterday?: number;
  time?: string;
  quick?: number;
  timezone?: string;
  fullDay?: boolean;
  noTimestamp?: boolean;
  noMetadata?: boolean;
  field?: string;
  comment?: string | string[];
  idPart?: string | string[];
  idDelimiter?: string;
  partition?: string;
  undo?: boolean;
  required?: string | string[];
  optional?: string | string[];
  remainder?: string;
  stringRemainder?: boolean;
  lenient?: boolean;
  showAll?: boolean;
  _?: (string | number)[];
};

export const configuredYargs = yargs
  .options({
    // couchdb options
    db: {
      describe: "The database to use, defaults to datum",
      alias: "database",
      nargs: 1,
    },
    host: {
      describe: "Host and port to use, defaults to 'localhost:5984'",
      nargs: 1,
    },
    username: {
      describe: "couchdb username to use'",
      nargs: 1,
    },
    password: {
      describe: "couchdb password to use'",
      nargs: 1,
    },
    env: {
      describe:
        "Environment file to read with COUCHDB_USER, COUCHDB_PASSWORD, COUCHDB_HOST",
      nargs: 1,
      normalize: true,
    },

    "base-data": {
      describe:
        "base object on which additional keys are added. Fed through relaxed-json, but should still parse to an object. Use with --no-metadata for raw json input into couchdb. Default: {}",
      nargs: 1,
      alias: "b",
      type: "string",
    },

    // timing
    date: {
      describe:
        "date of the timestamp, use `+n` or `-n` for a date relative to today. If no time is specified with -t, -T is assumed.",
      alias: "d",
      nargs: 1,
      type: "string",
    },
    yesterday: {
      describe:
        "use yesterday's date. Equivalent to `-d yesterday`. Use multiple times to go back more days",
      alias: "y",
      type: "count",
    },
    time: {
      describe:
        "specify time of the timestamp, use `+n` or `-n` for a timestamp n minutes relative to now",
      alias: "t",
      nargs: 1,
      type: "string",
    },
    quick: {
      describe:
        "quick options for time, use multiple times. -q = 5 min ago, -qq = 10 min ago, etc.",
      alias: "q",
      type: "count",
    },
    timezone: {
      describe:
        "Set the timezone to use instead of local time. Accepts both timezone names (America/Chicago) and utc offsets '-7'",
      alias: "z",
      type: "string",
    },
    "full-day": {
      describe:
        "make an entry for the full day, without a specific timestamp, occurs also when -d is used without -t",
      alias: "D",
      type: "boolean",
    },
    "no-timestamp": {
      describe: "omit the occurTime from the data",
      alias: "T",
      type: "boolean",
    },
    "no-metadata": {
      describe: "do not include meta data in document",
      alias: "M",
      type: "boolean",
    },
    // data
    field: {
      describe:
        "field specifying what is being tracked, used by default as partition for the data, but can be changed with --partition",
      alias: "f",
      nargs: 1,
      type: "string",
    },
    comment: {
      describe: "comment to include in the data",
      alias: "c",
      nargs: 1,
      type: "string",
    },

    // id
    "id-part": {
      describe:
        "Which field(s) to use for the _id field in the document." +
        " Can either be a single string with fields delimited by --id-delimiter" +
        " or can be used multiple times to progressively assemble an id delimited by --id-delimiter",
      alias: ["id", "pk", "_id"],
      type: "string",
    },
    "id-delimiter": {
      describe: "spacer between fields in the id",
      type: "string",
    },
    partition: {
      describe:
        "field to use for the partition (default: field, specified with -f)." +
        " Can be fields of data or raw strings surrounded by single quotes." +
        " Like --id-field, can be used  multiple times to assemble a partition separated by --id-delimiter",
      type: "string",
    },
    undo: {
      describe: "undoes the last datum entry, can be combined with -f",
      alias: "u",
      type: "boolean",
    },
    // "force-undo": {
    //   describe:
    //     "forces an undo, even if the datapoint was entered more than 15 minutes ago",
    //   alias: "U",
    //   type: "boolean",
    // },
    required: {
      describe:
        "Add a required key to the data, will be filled with first keyless data. If not enough data is specified to fill all required keys, an error will be thrown",
      alias: ["K", "req"],
      type: "string",
      nargs: 1,
    },
    optional: {
      describe:
        "Add an optional key to the data, will be filled with first keyless data. A default value can be specified with an '=', e.g., -k key=value",
      alias: ["k", "opt"],
      type: "string",
      nargs: 1,
    },
    remainder: {
      describe:
        "Any extra data supplied will be put into this key as an array. When --lenient is specified, defaults to 'extraData'",
      alias: ["rem", "R"],
      type: "string",
      nargs: 1,
    },
    "string-remainder": {
      describe:
        "Remainder data will be a space concatenated string rather than an array",
      alias: "S",
      type: "boolean",
    },
    lenient: {
      describe: "Allow extra data without defined keys",
      type: "boolean",
      alias: "l",
    },
    "show-all": {
      describe: "Show complete document when displaying, not just data",
      type: "boolean",
      alias: "A",
    },
  })
  .help("h")
  .alias("h", "help");
