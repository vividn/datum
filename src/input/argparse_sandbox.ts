import { ArgumentParser } from "argparse";
import * as shellQuote from "shell-quote";
import { MainDatumArgs } from "./mainArgs";

export function argparse_sandbox(cliInput: string | string[]): MainDatumArgs {
  const input =
    typeof cliInput === "string"
      ? shellQuote.parse(cliInput).map(String)
      : cliInput;
  const parser = new ArgumentParser({
    description: "argparse sandbox",
  });

  parser.add_argument("--db", {
    help: "The database to use, defaults to datum",
    dest: "db",
  });
  parser.add_argument("--host", {
    help: "Host and port to use, defaults to 'localhost:5984'",
    dest: "host",
  });
  parser.add_argument("--adapter", {
    help: "PouchDb adapter to use, will default to PouchDBs choice",
    dest: "adapter",
  });
  parser.add_argument("--username", {
    help: "couchdb username to use",
    dest: "username",
  });
  parser.add_argument("--password", {
    help: "couchdb password to use",
    dest: "password",
  });
  parser.add_argument("--env", {
    help: "Environment file to read with COUCHDB_USER, COUCHDB_PASSWORD, COUCHDB_HOST",
    dest: "env",
  });
  parser.add_argument("--create-db", {
    help: "Create the db if it does not exist",
    action: "store_true",
    dest: "createDb",
  });

  return parser.parse_args(input);
}
