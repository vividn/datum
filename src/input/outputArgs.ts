import yargs, { Argv } from "yargs";

export type OutputArgs = {
  showAll?: boolean;
  show?: Show;
  formatString?: string;
};

export enum Show {
  Default = "default",
  None = "none",
  Minimal = "minimal",
  Standard = "standard",
  Format = "format",
  All = "all",
}

export function outputYargs(otherYargs?: Argv): Argv {
  const yarg = otherYargs ?? yargs;
  return yarg.group(["show", "showAll", "formatString"], "Output").options({
    "show-all": {
      describe: "Show complete document when displaying, not just data",
      type: "boolean",
      alias: "A",
    },
    show: {
      describe: "how much of documents to show",
      type: "string",
      choices: Object.values(Show),
      default: "default",
      conflict: "show-all",
    },
    "format-string": {
      describe:
        "create a custom output string for visualizing the doc(s). Specify %keys% with percent signs",
      type: "string",
      alias: "o",
    },
  });
}
