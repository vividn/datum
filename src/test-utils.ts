import Nano, { DocumentScope } from "nano";
import { CouchDbError } from "./errors";
import { EitherPayload } from "./documentControl/DatumDocument";
import * as connectDb from "./auth/connectDb";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "./time/parseTimeStr";
import { now } from "./time/timeUtils";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const testNano = Nano(`http://admin:password@localhost:5983`);

export const mockDocMissingError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing",
  error: "not_found",
  reason: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "deleted",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing_named_view",
  error: "not_found",
  reason: "missing_named_view",
};

export async function resetTestDb(
  dbName: string
): Promise<DocumentScope<EitherPayload>> {
  const maxTries = 3;
  let tries = 0;
  await testNano.db.destroy(dbName).catch(pass);
  while (tries++ <= maxTries) {
    await testNano.db.destroy(dbName).catch(pass);
    await testNano.db.create(dbName).catch(pass);
    if ((await testNano.db.list()).includes(dbName)) {
      return testNano.use(dbName) as DocumentScope<EitherPayload>;
    }
  }
  throw Error(`Unable to reset database after ${maxTries} attempts`);
}

export function testDbLifecycle(dbName: string): DocumentScope<EitherPayload> {
  const db = testNano.use(dbName) as DocumentScope<EitherPayload>;

  beforeEach(async () => {
    await resetTestDb(dbName);
    jest.spyOn(connectDb, "connectDb").mockImplementation(async () => db);
  });

  afterEach(async () => {
    await testNano.db.destroy(dbName).catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => O
): (...args: A) => O {
  return (...args: A): O => {
    pushNow(timeStr);
    const returnVal: O = fn(...args);
    popNow();
    return returnVal;
  };
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
