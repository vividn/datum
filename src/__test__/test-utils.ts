import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
