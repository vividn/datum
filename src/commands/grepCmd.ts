import { MainDatumArgs } from "../input/mainArgs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { connectDb } from "../auth/connectDb";
import { showExists } from "../output/output";
import { updateLastDocsRef } from "../documentControl/lastDocs";
import { outputArgs } from "../input/outputArgs";
import { dbArgs } from "../input/dbArgs";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const grepArgs = new ArgumentParser({
  add_help: false,
});
grepArgs.add_argument("patterns", {
  help: "pattern to match",
  nargs: "+",
});

export const grepCmdArgs = new ArgumentParser({
  description: "Find all docs that match pattern in the database",
  prog: "datum grep",
  usage: "%(prog)s <patterns..>",
  parents: [grepArgs, dbArgs, outputArgs],
});

export type GrepCmdArgs = MainDatumArgs & {
  patterns: string[];
};

export async function grepCmd(
  args: GrepCmdArgs | string | string[],
  preparsed?: Partial<GrepCmdArgs>,
): Promise<EitherDocument[]> {
  const parsed = parseIfNeeded(grepCmdArgs, args, preparsed);
  const db = connectDb(parsed);
  const allDocs = (await db.allDocs({ include_docs: true })).rows;
  const matchingDocs: EitherDocument[] = allDocs.reduce(
    (accum: EitherDocument[], row) => {
      const docString = JSON.stringify(row.doc);
      if (parsed.patterns.every((pattern) => docString.match(pattern))) {
        accum.push(row.doc as EitherDocument);
      }
      return accum;
    },
    [],
  );
  const ids = matchingDocs.map((doc) => {
    showExists(doc, parsed);
    return doc._id;
  });
  await updateLastDocsRef(db, ids);
  return matchingDocs;
}
