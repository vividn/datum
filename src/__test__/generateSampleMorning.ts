import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
