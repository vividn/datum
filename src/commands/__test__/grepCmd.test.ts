import { mockedLogLifecycle, testDbLifecycle } from "../../__test__/test-utils";
import { grepCmd } from "../grepCmd";
import { Show } from "../../input/outputArgs";

describe("grepCmd", () => {
  const dbName = "grep_cmd_test";
  const db = testDbLifecycle(dbName);
  const mockedLog = mockedLogLifecycle();

  it("finds all the docs that contain the given string", async () => {
    const doc1 = { _id: "doc1", foo: "bar" };
    const doc2 = { _id: "doc2", data: { foo: "baz" } };
    const doc3 = { _id: "doc3", data: { foo: "bbarr" } };
    const doc4 = { _id: "doc4", data: { foo: "qux" } };
    const doc5 = { _id: "doc5", data: { bar: "foo" } };
    const doc6 = {
      _id: "doc6",
      data: { abc: "def" },
      meta: { humanId: "bar" },
    };
    await db.bulkDocs([doc1, doc2, doc3, doc4, doc5, doc6]);

    const retDocs = await grepCmd({
      db: dbName,
      patterns: ["bar"],
      show: Show.Standard,
    });
    expect(retDocs).toHaveLength(4);
    expect(retDocs).toEqual(
      expect.arrayContaining([
        expect.objectContaining(doc1),
        expect.objectContaining(doc3),
        expect.objectContaining(doc5),
        expect.objectContaining(doc6),
      ]),
    );

    expect(mockedLog.mock.calls).toMatchSnapshot();
  });

  it.todo("can exclude docs with match an exclude pattern");
  it.todo("can match based on matching multiple regex patterns");
  it.todo("can match based on any given regex pattern with the --or option ");
});
