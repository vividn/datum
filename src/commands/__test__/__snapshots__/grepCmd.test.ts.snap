// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grepCmd finds all the docs that contain the given string 1`] = `
[
  [
    "[33mEXISTS[39m [90mdoc1[39m",
  ],
  [
    "  foo: "bar"  ",
  ],
  [
    "[33mEXISTS[39m [90mdoc3[39m",
  ],
  [
    "  data: { foo: "bbarr" }  ",
  ],
  [
    "[33mEXISTS[39m [90mdoc5[39m",
  ],
  [
    "  data: { bar: "foo" }  ",
  ],
  [
    "[33mEXISTS[39m [33mbar[39m [90mdoc6[39m",
  ],
  [
    "  abc: "def"  ",
  ],
]
`;
